### ISSUE TITLE
`beginning-of-defun` doesn't work properly when narrowed

### ISSUE THREAD

**Issue by leungbk** [greghendershott/racket-mode#694]
Given an `init.el` that looks like

```elisp
(let ((bootstrap-file (concat user-emacs-directory "straight/repos/straight.el/bootstrap.el"))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(straight-use-package 'use-package)
(setq straight-use-package-by-default t)

(use-package racket-mode)
```

and a Racket file 

```racket
#lang racket

(define (hello _)
  42)

(define (bye-bye _)
  42)
```

I find that when I `C-x n d` within either function and try to use `beginning-of-defun`, `point` does not move where I expect it to.

<details>
<h2>Package</h2>
<dl>
<dt>metadata</dt><dd><pre>nil
</pre></dd>
<dt>package-archives</dt><dd><pre>(("gnu" . "https://elpa.gnu.org/packages/")
 ("nongnu" . "https://elpa.nongnu.org/nongnu/"))
</pre></dd>
<dt>racket--el-source-dir</dt><dd><pre>"/home/brian/.config/emacs/straight/build/racket-mode/"
</pre></dd>
<dt>racket--rkt-source-dir</dt><dd><pre>"/home/brian/.config/emacs/straight/build/racket-mode/racket/"
</pre></dd>
</dl>
<h2>System values</h2>
<dl>
<dt>emacs-version</dt><dd><pre>"30.0.50"
</pre></dd>
<dt>major-mode</dt><dd><pre>help-mode
</pre></dd>
<dt>system-type</dt><dd><pre>gnu/linux
</pre></dd>
<dt>x-gtk-use-system-tooltips</dt><dd><pre>t
</pre></dd>
<dt>display-graphic-p</dt><dd><pre>nil
</pre></dd>
</dl>
<h2>Buffer values</h2>
<dl>
<dt>after-change-functions</dt><dd><pre>nil
</pre></dd>
<dt>before-change-functions</dt><dd><pre>nil
</pre></dd>
<dt>completion-at-point-functions</dt><dd><pre>(tags-completion-at-point-function)
</pre></dd>
<dt>eldoc-documentation-function</dt><dd><pre>eldoc-documentation-default
</pre></dd>
<dt>font-lock-defaults</dt><dd><pre>nil
</pre></dd>
<dt>pre-command-hook</dt><dd><pre>(tooltip-hide)
</pre></dd>
<dt>post-command-hook</dt><dd><pre>(global-font-lock-mode-check-buffers global-eldoc-mode-check-buffers
				     mode-local-post-major-mode-change)
</pre></dd>
<dt>post-self-insert-hook</dt><dd><pre>(electric-indent-post-self-insert-function
 blink-paren-post-self-insert-function)
</pre></dd>
<dt>xref-backend-functions</dt><dd><pre>(etags--xref-backend)
</pre></dd>
</dl>
<h2>Racket Mode values</h2>
<dl>
<dt>racket--cmd-open-p</dt><dd><pre>nil
</pre></dd>
<dt>racket-after-run-hook</dt><dd><pre>nil
</pre></dd>
<dt>racket-back-end-configurations</dt><dd><pre>((:directory "/" :racket-program nil :remote-source-dir nil
	     :restart-watch-directories nil :windows nil))
</pre></dd>
<dt>racket-before-run-hook</dt><dd><pre>(racket-ansi-color-context-reset)
</pre></dd>
<dt>racket-browse-url-function</dt><dd><pre>racket-browse-url-using-temporary-file
</pre></dd>
<dt>racket-command-timeout</dt><dd><pre>10
</pre></dd>
<dt>racket-documentation-search-location</dt><dd><pre>"https://docs.racket-lang.org/search/index.html?q=%s"
</pre></dd>
<dt>racket-error-context</dt><dd><pre>medium
</pre></dd>
<dt>racket-hash-lang-token-face-alist</dt><dd><pre>((constant . font-lock-constant-face) (error . error)
 (other . font-lock-doc-face) (keyword . font-lock-keyword-face)
 (hash-colon-keyword . racket-keyword-argument-face)
 (at . font-lock-doc-face))
</pre></dd>
<dt>racket-history-filter-regexp</dt><dd><pre>"\\`\\s *\\'"
</pre></dd>
<dt>racket-imagemagick-props</dt><dd><pre>nil
</pre></dd>
<dt>racket-images-inline</dt><dd><pre>t
</pre></dd>
<dt>racket-images-keep-last</dt><dd><pre>100
</pre></dd>
<dt>racket-images-system-viewer</dt><dd><pre>"display"
</pre></dd>
<dt>racket-indent-curly-as-sequence</dt><dd><pre>t
</pre></dd>
<dt>racket-indent-sequence-depth</dt><dd><pre>0
</pre></dd>
<dt>racket-logger-config</dt><dd><pre>((cm-accomplice . warning) (GC . info) (module-prefetch . warning)
 (optimizer . info) (racket/contract . error)
 (racket-mode-debugger . info) (sequence-specialization . info)
 (* . fatal))
</pre></dd>
<dt>racket-memory-limit</dt><dd><pre>2048
</pre></dd>
<dt>racket-mode-hook</dt><dd><pre>nil
</pre></dd>
<dt>racket-module-forms</dt><dd><pre>"\\s(\\(?:module[*+]?\\|library\\)"
</pre></dd>
<dt>racket-pretty-lambda</dt><dd><pre>nil
</pre></dd>
<dt>racket-pretty-print</dt><dd><pre>t
</pre></dd>
<dt>racket-program</dt><dd><pre>"racket"
</pre></dd>
<dt>racket-repl-buffer-name-function</dt><dd><pre>nil
</pre></dd>
<dt>racket-repl-command-file</dt><dd><pre>"/home/brian/.config/emacs/racket-mode/repl.rkt"
</pre></dd>
<dt>racket-repl-history-directory</dt><dd><pre>"~/.config/emacs/racket-mode/"
</pre></dd>
<dt>racket-repl-mode-hook</dt><dd><pre>nil
</pre></dd>
<dt>racket-sexp-comment-fade</dt><dd><pre>0.5
</pre></dd>
<dt>racket-shell-or-terminal-function</dt><dd><pre>racket-shell
</pre></dd>
<dt>racket-show-functions</dt><dd><pre>(racket-show-pseudo-tooltip)
</pre></dd>
<dt>racket-smart-open-bracket-enable</dt><dd><pre>nil
</pre></dd>
<dt>racket-submodules-to-run</dt><dd><pre>((test) (main))
</pre></dd>
<dt>racket-use-repl-submit-predicate</dt><dd><pre>nil
</pre></dd>
<dt>racket-xp-add-binding-faces</dt><dd><pre>nil
</pre></dd>
<dt>racket-xp-after-change-refresh-delay</dt><dd><pre>1
</pre></dd>
<dt>racket-xp-highlight-unused-regexp</dt><dd><pre>"^[^_]"
</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval (racket--xp-mode-lighter))
</pre></dd>
</dl>
<h2>Minor modes</h2>
<dl>
<dt>enabled</dt><dd><pre>((auto-composition-mode) (auto-compression-mode)
 (auto-encryption-mode) (blink-cursor-mode) (buffer-read-only)
 (electric-indent-mode) (file-name-shadow-mode) (font-lock-mode)
 (global-eldoc-mode) (global-font-lock-mode) (gpm-mouse-mode)
 (indent-tabs-mode) (isearch-fold-quotes-mode) (line-number-mode)
 (menu-bar-mode) (minibuffer-regexp-mode) (mouse-wheel-mode)
 (semantic-minor-modes-format) (show-paren-mode)
 (straight-package-neutering-mode) (straight-use-package-mode)
 (tool-bar-mode) (tooltip-mode) (transient-mark-mode))
</pre></dd>
<details><summary>Disabled minor modes</summary>
<dt>disabled</dt><dd><pre>((abbrev-mode) (auto-fill-function) (auto-fill-mode) (auto-save-mode)
 (auto-save-visited-mode) (buffer-face-mode) (button-mode)
 (cl-old-struct-compat-mode) (column-number-mode)
 (comint-fontify-input-mode) (compilation-minor-mode)
 (compilation-shell-minor-mode) (completion-in-region-mode)
 (context-menu-mode) (cursor-face-highlight-mode) (defining-kbd-macro)
 (eldoc-mode) (electric-layout-mode) (electric-quote-mode)
 (global-prettify-symbols-mode) (global-semantic-highlight-edits-mode)
 (global-semantic-highlight-func-mode)
 (global-semantic-show-parser-state-mode)
 (global-semantic-show-unmatched-syntax-mode)
 (global-semantic-stickyfunc-mode) (global-visual-line-mode)
 (horizontal-scroll-bar-mode) (hs-minor-mode) (isearch-mode)
 (jit-lock-debug-mode) (lock-file-mode) (lost-selection-mode)
 (modifier-bar-mode) (next-error-follow-minor-mode) (overwrite-mode)
 (paragraph-indent-minor-mode) (prettify-symbols-mode)
 (racket-smart-open-bracket-mode) (racket-xp-mode)
 (read-extended-command-mode) (semantic-highlight-edits-mode)
 (semantic-highlight-func-mode) (semantic-mode)
 (semantic-show-parser-state-mode)
 (semantic-show-unmatched-syntax-mode) (semantic-stickyfunc-mode)
 (sh-electric-here-document-mode) (shell-dirtrack-mode)
 (shell-highlight-undef-mode) (size-indication-mode)
 (straight-live-modifications-mode) (straight-symlink-emulation-mode)
 (tab-bar-history-mode) (tab-bar-mode) (temp-buffer-resize-mode)
 (text-scale-mode) (treesit-explore-mode) (treesit-inspect-mode)
 (undelete-frame-mode) (url-handler-mode) (use-hard-newlines)
 (view-mode) (visible-mode) (visual-line-mode) (window-divider-mode)
 (xref-etags-mode))
</pre></dd>
</details>
</dl>
</details>

**Comment by leungbk** [comment]
Some other unexpected behavior I noticed: after running `C-c C-c` on the buffer, when the buffer is narrowed to either function, `C-M-x` does not work when `point` is on the first line of either `define`; however, when on the second line of either function, `C-M-x` works fine.

**Comment by greghendershott** [comment]
Hmm. Unlike #693, I can't seem to reproduce this, so far.  For me:

- <kbd>C-M-a</kbd> always moves point before the open paren of the definition.
- <kbd>C-M-x</kbd> does work on either line.

I'll try with Emacs built from source, later.

**Comment by leungbk** [comment]
Thanks for your reply!

> Hmm. Unlike #693, I can't seem to reproduce this, so far. For me:
> 
>     * C-M-a always moves point before the open paren of the definition.

Hmm, I can't seem to get the expected behavior; I also tried on Emacs 28.2 and 29.1 just now and saw no change.

>     * C-M-x does work on either line.

After I `C-c C-c`, running `narrow-to-defun` followed by `C-M-x` when `point` is anywhere on the first line of the second `define` yields this in the `*Racket REPL*`:

```
Welcome to Racket v8.11.1 [cs].
————— run hello.rkt —————
hello.rkt> 
racket-mode-repl::10: _: wildcard not allowed as an expression
  in: _
hello.rkt> 
```

Feel free to let me know if there's any other info I can provide.

**Comment by greghendershott** [comment]
I tried again with Emacs built from source, c. couple weeks ago (`GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.16.0) of 2023-12-07`). But same result: still can't reproduce. :(

FWIW I run it with `~/src/emacs/src/emacs -L ~/src/elisp/racket-mode/ -Q` and manually `M-x load-file` the `racket-mode.el` file.

Then I visit `issue-694.rkt` file (your example above). `racket-mode` is enabled automatically. I try your steps. Everything works as expected.

---

Even with `-Q` my enabled minor modes are:

```
((auto-composition-mode) (auto-compression-mode)
 (auto-encryption-mode) (blink-cursor-mode) (buffer-read-only)
 (electric-indent-mode) (file-name-shadow-mode) (font-lock-mode)
 (global-eldoc-mode) (global-font-lock-mode) (indent-tabs-mode)
 (isearch-fold-quotes-mode) (line-number-mode) (menu-bar-mode)
 (minibuffer-regexp-mode) (mouse-wheel-mode)
 (semantic-minor-modes-format) (show-paren-mode) (tool-bar-mode)
 (tooltip-mode) (transient-mark-mode))
```

Yours are:


```
((auto-composition-mode) (auto-compression-mode)
 (auto-encryption-mode) (blink-cursor-mode) (buffer-read-only)
 (electric-indent-mode) (file-name-shadow-mode) (font-lock-mode)
 (global-eldoc-mode) (global-font-lock-mode) (gpm-mouse-mode)
 (indent-tabs-mode) (isearch-fold-quotes-mode) (line-number-mode)
 (menu-bar-mode) (minibuffer-regexp-mode) (mouse-wheel-mode)
 (semantic-minor-modes-format) (show-paren-mode)
 (straight-package-neutering-mode) (straight-use-package-mode)
 (tool-bar-mode) (tooltip-mode) (transient-mark-mode))
```

Those are nearly identical -- if I'm eyeballing it correctly, you have `gpm-mouse-mode` enabled which I don't -- so I am super stumped...

(This feels like one of those bugs that will turn out to be either amazingly subtle, or, blatantly obvious/simple in hindsight, but nothing in between.)

**Comment by greghendershott** [comment]
p.s. FWIW I build with the following on Debian 12 (e.g. no libjit ergo no native-comp):

```sh
## Clone
git clone git://git.savannah.gnu.org/emacs.git

## Deps
sudo apt install \
  autoconf \
  build-essential \
  libgtk-3-dev \
  libgnutls28-dev \
  libtiff5-dev \
  libgif-dev \
  libjpeg-dev \
  libpng-dev \
  libxpm-dev \
  libxml2-dev \
  libncurses-dev \
  texinfo

## Build
./autogen.sh
./configure
make --jobs=$(nproc)
```

**Comment by greghendershott** [comment]
> blatantly obvious/simple in hindsight

This. I had a tentative fix for #693. I thought I had reverted that, while trying to reproduce this -- but I hadn't really.

That fix for #693 turns out to fix this, too, in a way I didn't expect.

I am so sorry for the noise!

I'll push a commit to fix both issues... but since I'm being so dumb today, I'll slow down and triple check things...
